### PR TITLE
Package manager extension loads extensions and view extensions

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

https://jira.autodesk.com/browse/QNTM-3232

This PR starts to add extension loading to package manager.
This is done in the following way:

* The `PackageLoader` looks for extension and viewExtension manifest files while loading packages and raises some events requesting the extensions be loaded and added.

    The `PackageManagerExtension` forwards these events on to the`ExtensionManager` and `ExtensionLoader` or the View versions of those classes.

* Tests are added to the PackageLoader for testing late loading, and some new tests are added which test packages being loaded from startup which include extensions and viewExtensions. These sample extensions are part of the DynamoSamples repo (I will send a PR there as well for a small update)

* TODO
- [x] more rigorous manual testing
- [x] test uploading and downloading packages to packageManager (right now only local packages are tested)
- [x] Since late loaded packages do not get a `Startup()` call, I think it makes sense to have `ReadyParams` have a property which is a `StartupParams` object. This way even extensions loaded this way will not miss references to things like` Dynamo version` and `CustomNodeManager`.
- [ ] you tell me :)

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers
@ramramps 
@pboyer - would be good to get your thoughts Peter if you have some time.

### FYIs

@alfarok @Racel

### CODE REVIEW

File: test/DynamoCoreWpfTests/DynamoTestUIBase.cs [hunk 1]
```diff
@@ -83,6 +78,23 @@ public virtual void Start()
             SynchronizationContext.SetSynchronizationContext(new SynchronizationContext());
         }
 
+        /// <summary>
+        /// Derived test classes could override it to provide different 
+        /// configuration.
+        /// </summary>
+        /// <param name="settings"></param>
+        /// <returns></returns>
```
Comment (pboyer) [inline test/DynamoCoreWpfTests/DynamoTestUIBase.cs]: Don't include the `params` and `returns` object at all if they're empty. Make the summary a complete sentence.

File: test/DynamoCoreWpfTests/PackageManagerExtensionLoadingTests.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Controls;
+
+namespace DynamoCoreWpfTests
+{
+    class PackageManagerExtensionLoadingTests : DynamoTestUIBase
+    {
+        public string PackagesDirectory { get { return Path.Combine(GetTestDirectory(this.ExecutingDirectory), "pkgs"); } }
```
Comment (pboyer) [inline test/DynamoCoreWpfTests/PackageManagerExtensionLoadingTests.cs]: Can this be a `private static readonly string`?

File: test/DynamoCoreWpfTests/PackageManagerExtensionLoadingTests.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+       // var loader = GetPackageLoader();
+       // var pkgDir = Path.Combine(PackagesDirectory, "SampleExtension");
```
Comment (pboyer) [inline test/DynamoCoreWpfTests/PackageManagerExtensionLoadingTests.cs]: Delete dead code.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 4]
```diff
@@ -749,6 +775,11 @@ protected DynamoModel(IStartConfiguration config)
             StartBackupFilesTimer();
 
             TraceReconciliationProcessor = this;
+            //we're done loading the model
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: This comment is self-evident. Maybe you want to write something like `This event should only be raised when at the end of this method`.

File: src/DynamoCore/Extensions/ExtensionManager.cs [hunk 5]
```diff
@@ -8,6 +8,40 @@
...
+
+    /// <summary>
+    /// An object which may request extensions to be loaded and added to the extensionsManager
+    /// </summary>
+    public interface IExtensionSource
+    {
+        /// <summary>
+        /// event that is raised when package manager requests an extension be loaded
+        /// which was found within a package it is currently loading.
+        /// </summary>
```
Comment (pboyer) [inline src/DynamoCore/Extensions/ExtensionManager.cs]: This `event` is more general than the comment suggests. No?

Is it idiomatic to suffix an `event` with `Handler`? https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-events

File: src/DynamoCore/Models/DynamoModel.cs [hunk 6]
```diff
@@ -198,6 +198,12 @@ private void OnShutdownCompleted()
                 ShutdownCompleted(this);
         }
 
+         /// <summary>
+         /// this event is raised when Dynamo is ready for user interaction.
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: The summary should be a complete sentence with appropriate capitalization.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 7]
```diff
@@ -668,6 +674,24 @@ protected DynamoModel(IStartConfiguration config)
             extensionManager.MessageLogged += LogMessage;
             var extensions = config.Extensions ?? LoadExtensions();
 
+            //when dynamo is ready, alert the loaded extensions
```
Comment (pboyer) [inline src/DynamoCore/Models/DynamoModel.cs]: A minor point of annoyance. Please add a space between your `//` and the text. This is most definitely the style employed in this code base.

File: src/DynamoCore/Extensions/ExtensionManager.cs [hunk 8]
```diff
@@ -23,6 +57,29 @@ public class ExtensionManager: IExtensionManager, ILogSource
...
+        private void removeEventSubscriptions(IExtension obj)
+        {
+            if (obj is IExtensionSource)
+            {
+                (obj as IExtensionSource).RequestLoadExtensionHandler -= this.ExtensionLoader.Load;
+                (obj as IExtensionSource).RequestAddExtensionHandler -= this.Add;
+            }
+        }
+
+        private void AddEventSubscriptions(IExtension obj)
```
Comment (pboyer) [inline src/DynamoCore/Extensions/ExtensionManager.cs]: `SubscribeExtension` & `UnsubscribeExtension` is probably a more evocative name.

File: src/DynamoCore/Extensions/ExtensionManager.cs [hunk 9]
```diff
@@ -8,6 +8,40 @@
...
+        /// which was found within a package it is currently loading.
+        /// </summary>
+        event Func<string, dynamic> RequestLoadViewExtension;
+
+        /// <summary>
+        /// event that is raised when package manager requests a viewExtension to be added to
+        /// the list of currently loaded extensions. The ViewExtension was found within a package that
+        /// is currently being loaded. This is usually raised directly after a ViewExtension load occurs.
+        /// </summary>
+        event Action<dynamic> RequestAddViewExtension;
```
Comment (pboyer) [inline src/DynamoCore/Extensions/ExtensionManager.cs]: All of these events have an unnecessarily large level of focus on the package manager.

Comment (pboyer) [reply src/DynamoCore/Extensions/ExtensionManager.cs]: `ViewExtensions` aren't really part of `DynamoCore`. Shouldn't this be in the `DynamoCoreWpf` layer? It might seem duplicative, but I don't think this really results in that much extra code.

Comment (mjkkirschner) [reply src/DynamoCore/Extensions/ExtensionManager.cs]: indeed thats what I intended to do first, but I don't believe there is a View `packageLoader` analog where we would load extension manifests from package from (again) - since these are already loaded in the package manager extension (not a view extension) - we would not be able to reference an interface in `DynamoCoreWpf` without kind of entailing an extension with wpf references... so I decided to use dynamic here and add these events.

I could definitely be convinced to look for an alternative - any thoughts on that?

**Comment by mjkkirschner** [comment]
@pboyer thanks for the review - I will address all the documentation comments - and am interested in your thoughts on the DynamoCore /DynamoCoreWPF separation issue.

**Comment by mjkkirschner** [comment]
an update here - mostly I think things are resolved, but I noticed that there is an asymmetry between extensions and view extensions which are included in packages that are present from startup. namely - Extensions will NOT have their `startup` method called while viewExtensions will... this is not great, I am trying to find a solution, but I am not finding anything that will let me keep all other loading sequences the same as they are today... yet.

File: src/DynamoCore/Models/DynamoModel.cs [hunk 10]
```diff
@@ -749,6 +775,11 @@ protected DynamoModel(IStartConfiguration config)
             StartBackupFilesTimer();
 
             TraceReconciliationProcessor = this;
+            // This event should only be raised at the end of this method.
+             DynamoReady(new ReadyParams(this));
+        }
+
+        private void DynamoReadyExtensionHandler(ReadyParams readyParams, IEnumerable<IExtension> extensions) {
 
             foreach (var ext in ExtensionManager.Extensions)
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: this is a bug - we should use the passed extensions - will be fixed.

**Comment by mjkkirschner** [comment]
@ramramps @pboyer I think this is ready for another look.
I am still working to add more automated testing, but I have resolved the issue with extensions and ViewExtensions which are included in packages at startup - these will both now have their `startup` methods called. extensions loaded later will not - this reflects what is stated in the documentation.

not sure whats up with the mono build.

also please note that I updated the `ReadyParams` and `ViewLoadedParams` to with a reference to the `StartupParams` as I noted I might do above.

File: src/DynamoCore/Extensions/ReadyParams.cs [hunk 11]
```diff
@@ -28,6 +28,22 @@ internal ReadyParams(DynamoModel dynamoM)
             dynamoM.Logger.NotificationLogged += OnNotificationRecieved;
         }
 
+        /// <summary>
+        /// A reference to the <see cref="StartupParams"/> class.
+        /// Useful if this extension will be loaded from a package as its startup method, will not be called.
+        /// </summary>
```
Comment (ramramps) [inline src/DynamoCore/Extensions/ReadyParams.cs]: may be a wrong comma here?

File: src/DynamoCore/Extensions/ExtensionLoader.cs [hunk 12]
```diff
@@ -122,5 +124,10 @@ private void Log(string msg)
         {
             Log(LogMessage.Info(msg));
         }
+
+        /// <summary>
+        /// An event that is raised when an extension starts loading.
+        /// </summary>
+        public event Action<IExtension> ExtensionLoading;
```
Comment (pboyer) [inline src/DynamoCore/Extensions/ExtensionLoader.cs]: I think standard practice is to name an `event` with a verb. I know this seems dumb, but having conventions make it much easier to work on a code base.

Comment (mjkkirschner) [reply src/DynamoCore/Extensions/ExtensionLoader.cs]: you mean with the verb first?
`LoadingExtension`?

Comment (pboyer) [reply src/DynamoCore/Extensions/ExtensionLoader.cs]: Eh, forget it. It's probably good.

File: src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs [hunk 13]
```diff
@@ -23,6 +23,19 @@ public class ViewLoadedParams : ReadyParams
...
+        /// <summary>
+        /// A reference to the <see cref="ViewStartupParams"/> class.
+        /// Useful if this extension will be loaded from a package as its startup method, will not be called.
+        /// </summary>
+        public ViewStartupParams ViewStartupParameters
+        {
+            get
+            {
+                var startupParams = new ViewStartupParams(this.dynamoViewModel);
+                return startupParams;
```
Comment (pboyer) [inline src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs]: Why not `return new ViewStartupParams(this.dynamoViewModel);`

Comment (pboyer) [inline src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs]: I would recommend making this simply a `private readonly` field with a `property`. Also, why not name this the same as the type name - `ViewStartupParams`?

File: src/DynamoPackages/PackageLoader.cs [hunk 14]
```diff
@@ -151,6 +159,21 @@ public void Load(Package package)
...
+                // If the additional files contained an extension or extensionView manifest then request it be loaded.
+                var extensionManifests = package.AdditionalFiles.Where(
+                    file => file.Model.Name.Contains("ExtensionDefinition.xml")).ToList();
+                foreach (var extPath in extensionManifests)
+                {
+                    var extension = RequestLoadExtension?.Invoke(extPath.Model.FullName);
+                    if (extension != null)
+                    {
+                        RequestAddExtension?.Invoke(extension);
+
```
Comment (pboyer) [inline src/DynamoPackages/PackageLoader.cs]: Anomalous white space?

File: src/DynamoPackages/PackageLoader.cs [hunk 15]
```diff
@@ -29,8 +30,15 @@ public class PackageLoader : LogSourceBase
...
         internal event Func<string, IEnumerable<CustomNodeInfo>> RequestLoadCustomNodeDirectory;
+        internal event Func<string, dynamic> RequestLoadExtension;
+        internal event Action<dynamic> RequestAddExtension;
         public event Action<Package> PackageAdded;
         public event Action<Package> PackageRemoved;
+        /// <summary>
+        /// A map of the extension manifest paths which the package loader found during package load
+        /// and has requested be loaded. The result of the load request is mapped to the requested path.
+        /// </summary>
+        public Dictionary<string, dynamic> RequestedExtensions { get; private set; } = new Dictionary<string, dynamic>();
```
Comment (pboyer) [inline src/DynamoPackages/PackageLoader.cs]: Can this be `internal`?

Comment (mjkkirschner) [reply src/DynamoPackages/PackageLoader.cs]: if we expose `packageManagerExtension` internals to dynamocore

File: src/DynamoPackages/PackageManagerExtension.cs [hunk 16]
```diff
@@ -34,6 +38,11 @@ public string UniqueId
         /// </summary>
         public PackageLoader PackageLoader { get; private set; }
 
+        /// <summary>
+        ///   
+        /// </summary>
```
Comment (pboyer) [inline src/DynamoPackages/PackageManagerExtension.cs]: Looks like we don't need a comment here? :)

File: src/DynamoPackages/PackageManagerExtension.cs [hunk 17]
```diff
@@ -57,6 +66,16 @@ public void Dispose()
...
+            if (RequestLoadExtension != null)
+            {
+                PackageLoader.RequestLoadExtension -=
+                RequestLoadExtension;
+            }
+            if (RequestAddExtension != null)
+            {
+                PackageLoader.RequestAddExtension -=
+                RequestAddExtension;
+            }
```
Comment (pboyer) [inline src/DynamoPackages/PackageManagerExtension.cs]: Really strange whitespace here. Possibly an editing defect?

File: src/DynamoPackages/PackageLoader.cs (hunk 15 above)
Comment (mjkkirschner) [reply src/DynamoPackages/PackageLoader.cs]: actually I think we can just make it internal.

File: src/DynamoPackages/PackageManagerExtension.cs (hunk 16 above)
Comment (mjkkirschner) [reply src/DynamoPackages/PackageManagerExtension.cs]: nice catch thanks.

File: src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs (hunk 13 above)
Comment (mjkkirschner) [reply src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs]: will do.

File: src/DynamoCoreWpf/Extensions/ViewExtensionManager.cs [hunk 18]
```diff
@@ -14,8 +15,49 @@ internal class ViewExtensionManager : IViewExtensionManager, ILogSource
...
+        {
+            //if this extension could be a source of other extensions (like packageManagerExtension) then
+            //lets handle those requests.
+            if (obj is IExtensionSource)
+            {
+                (obj as IExtensionSource).RequestLoadExtension += RequestLoadViewExtensionHandler;
+                (obj as IExtensionSource).RequestAddExtension += RequestAddViewExtensionHandler;
+            }
+
+        }
```
Comment (pboyer) [inline src/DynamoCoreWpf/Extensions/ViewExtensionManager.cs]: There's a lot of distracting white space issues in this file. Sometimes a method has an extra line, sometimes methods are separated by a new line, sometimes not. No one likes to talk about source code formatting.

It like just style, but it _does_ matter. It's harder to read the code in Github, I feel the need to comment on it (distracting from meaningful review comments), it's harder to write tooling, and ultimately can be solved with automation.

Comment (mjkkirschner) [reply src/DynamoCoreWpf/Extensions/ViewExtensionManager.cs]: will fix them up.

**Comment by mjkkirschner** [comment]
@ramramps lets take a look at this together tomorrow if you have time. I added a semi complex test, but I think it's worth it.

I will also update the description of the PR to better match what it does now.

I still have not found time to test on dev PM yet.

**Comment by Racel** [comment]
@mjkkirschner - Agreed. Let us not increase the scope of the PR. Can you please file a separate task to fix this bug. IMO, we do not need it for 2.0.

**Comment by ramramps** [comment]
@mjkkirschner Just to be clear, is it only for View extensions / extensions? and not for the usual packages, correct?

**Comment by mjkkirschner** [comment]
@ramramps yep, the issue is only when you first publish a package containing an extension which you already didnt have install - thats an edge case I think as most devs would have their package installed already so they can test it locally before deploying I would think.
